### PR TITLE
[Spot] Event callback for adding notification 

### PR DESCRIPTION
**Pull request by MaoZiming**
Issue #2000: A user would like to have a customizable event callback for the spot job, so that they can implement some notification when the spot job status changes. Example:
```
event_callback:
  send_slack_message $JOB_ID $JOB_STATUS
```

Allows the controller to execute a customizable bash script when the spot status changes. 

Example:
Register a Slack message web hook from [here](https://api.slack.com/apps?new_app=1).
```
event_callback: |
  echo 'this is an event callback script'
  curl -X POST -H 'Content-type: application/json' --data "{\"text\":\"JOB_ID: $SKYPILOT_JOB_ID, JOB_STATUS: $JOB_STATUS\"}" https://hooks.slack.com/services/...
```

The user will receive a Slack notification when the spot status changes.
image

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/spot/controller.py
Comment (Michaelvll) [inline sky/spot/controller.py:292]: How about we just set the environment variable in the subprocess call?
```
subprocess.run(..., env=dict(os.environ, JOB_ID=self._job_id, ...), ...)
```

File: sky/spot/controller.py
Comment (Michaelvll) [inline sky/spot/controller.py:296]: Maybe we can get rid of this, as the controller log is normally not a user-facing log, and we probably have enough loggings in it.
If a user would like to log something to the controller log, they can also specify the bash script with `echo hello world` in it.

File: sky/spot/controller.py
Comment (Michaelvll) [inline sky/spot/controller.py:299]: `subprocess.run` does not support running a bash script. Maybe we should call the `log_lib.run_bash_command_with_log` function (may need to slightly modify the function's argument to make the `job_id` and `job_owner` optional.

File: sky/spot/controller.py
Comment (Michaelvll) [inline sky/spot/controller.py:304]: An idea for this (better UX):
```
logger.info(f'=== START: event callback for {state!r} ===')
log_lib.run_bash_command_with_log(...)
logger.info(f'=== END: event callback for {state!r} ===')
```

File: sky/spot/controller.py
Comment (Michaelvll) [inline sky/spot/controller.py:298]: Do you think we can just let the `event_callback` to just take a bash script instead of having this prefix?

File: sky/task.py
Comment (Michaelvll) [inline sky/task.py:152]: This may not be needed if we are only taking a script for the `event_callback`

**Review by Michaelvll — COMMENTED** [review]
Thank you for adding this feature @MaoZiming! It would be very useful for the users. Left several comments. 

Additional comments:
1. We may want to call the event callback when the spot job starts recovering as well, according to the users feedback.
2. We may want to think of whether we should run the callback asynchronously, instead of in a blocking way. Otherwise, our control logic may be significantly delayed.

**Comment by MaoZiming** [comment]
@Michaelvll Incorporated all the comments. The basic skeleton for slack notification should be there.

File: sky/spot/spot_state.py
Comment (Michaelvll) [inline sky/spot/spot_state.py:474]: Weirdly, when I am trying this PR, the `STARTING`, `RUNNING`, `RECOVERING`, `RECOVERED` works normally, while this `CANCELLING`/`CANCELLED` state was not sent to my slack. Do you have an idea why this happens?
Fyi, I did the following:
1. `sky spot launch test.yaml`
2. manually kill the cluster on the console
3. the states were correctly sent 
4. `sky spot cancel -a`: the cancelling/cancelled states were not correctly sent.

File: sky/spot/spot_state.py
Comment (Michaelvll) [inline sky/spot/spot_state.py:473]: This might need to go within the condition `rows.rowcount > 0` above, as this function will be called even if the job is already failed. We only need to call the event callback when the job is actually being cancelled. (same for the `set_cancelled` function)

File: sky/spot/spot_state.py
Comment (Michaelvll) [inline sky/spot/spot_state.py:495]: ditto

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @MaoZiming! The slack messages were sent like a magic. It works well for `STARTING`/`RUNNING`/`RECOVERING`/`RECOVERED`/`SUCCEEDED`, but it seems like not working well with the `CANCELLING`/`CANCELLED`.

File: sky/spot/controller.py
Comment (MaoZiming) [reply sky/spot/controller.py:299]: Thanks! We already have JOB_STATUS (e.g. RECOVERING). What additional EVENT_TYPE are you thinking about?

File: sky/spot/spot_state.py
Comment (MaoZiming) [reply sky/spot/spot_state.py]: I believe this might be because event_callback is non-blocking. Spot tasks could be killed first before the callback finishes execution.

Comment (MaoZiming) [reply sky/spot/spot_state.py]: For `set_cancelling` outside of `SpotController`, I didn't pass in `event_callback_func`. 
https://github.com/skypilot-org/skypilot/blob/spot-event-callback/sky/spot/controller.py#L456

File: sky/spot/controller.py
Comment (Michaelvll) [reply sky/spot/controller.py]: I am thinking in the future, a normal job on serverful cluster will also have event callback, where the `EVENT_TYPE` will be `Job`, while the current type can be `Spot`. That is for discussion and we can probably leave it to the future.

Comment (MaoZiming) [reply sky/spot/controller.py]: Oh i see!

File: sky/spot/spot_state.py
Comment (Michaelvll) [reply sky/spot/spot_state.py]: Do we want to pass the event_callback_func to those as well? Otherwise, `sky spot cancel -a` will not send the message.

Comment (MaoZiming) [reply sky/spot/spot_state.py]: I am thinking about a clean way to do this since we need to move `event_callback_func` outside of `SpotController`. Do you know if is there an easy way to get `task` from `task_id` and `job_id`? It seems to require `dag_yaml` or `self._dag`

Comment (Michaelvll) [reply sky/spot/spot_state.py]: Ahh, as we did for the `_cleanup` function, we can retrieve the dag using the `day_yaml` as we did in `_cleanup` function https://github.com/skypilot-org/skypilot/blob/438f81599fd1728e5b8df6df7ca72e804bd6e1fd/sky/spot/controller.py#L430

However, we are not able to get some of the environment variable. We can probably print some log in the event callback saying which variables are not accessible, and we can wait for the user feedback to see if they want to have some of the variables in the CANCEL.

Comment (MaoZiming) [reply sky/spot/spot_state.py]: @Michaelvll Thanks for all the detailed feedback! The suggested changes should all be incorporated

**Review by Michaelvll — APPROVED** [review]
Thanks for this exciting PR @MaoZiming! The code looks good to me. Several small nits.

File: sky/spot/controller.py
Comment (MaoZiming) [reply sky/spot/controller.py:437]: Yep. I think that's better. Thanks!

**Comment by MaoZiming** [comment]
Thank you so much @Michaelvll! I have incorporated all the suggestions.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @MaoZiming! I just realized that the current way to pass the callback function can be a bit confusing. Do you think we can do something like the comments below?

File: sky/spot/spot_state.py
Comment (MaoZiming) [reply sky/spot/spot_state.py]: Sure! Just pushed

Comment (Michaelvll) [reply sky/spot/spot_state.py]: Feel free to push back if you find the current way is better. I am thinking if we let the inner function take the argument `status`, we only need to create the callback function once inside the controller class, and the status can be passed in by those `set_*` functions. Do you think that would make the code cleaner? :)

For the other parts, I think the code is good to go if the tests are all passed. Thanks for the great effort on this @MaoZiming!

Comment (MaoZiming) [reply sky/spot/spot_state.py]: Thanks. I moved `job_state` to the inner function. 
One thing is that for `def run(self):`, we still have to instantiate `spot_utils.event_callback_func` every time `set_*` is called during the error handling, as `task_id` and `task` might change.

**Review by Michaelvll — APPROVED** [review]
Thanks for awesome work @MaoZiming! LGTM. Please feel free to merge it.

**Comment by concretevitamin** [comment]
Very nice @MaoZiming!

**Pull request merged**
